wgpu-core/src/command/compute.rs

```diff
@@ -851,6 +889,211 @@ impl Global {
+            match command {
+                ArcComputeCommand::SetBindGroup {
+                    index: 0,
+                    num_dynamic_offsets,
+                    ref bind_group,
+                } => {
+                    current_bind_group_0 = Some((num_dynamic_offsets, bind_group.clone()));
+                    new_commands.push(command);
+                }
+                ArcComputeCommand::SetPipeline(ref pipeline) => {
+                    current_pipeline = Some(pipeline.clone());
+                    new_commands.push(command);
+                }
+                ArcComputeCommand::DispatchIndirect { ref buffer, offset } => {
+                    // if there is no pipeline set, don't inject the validation commands as we will error anyway
+                    if let Some(original_pipeline) = current_pipeline.clone() {
+                        // validate some buffer properties that won't be validated later
+                        check_buffer_usage(
+                            buffer.as_info().id(),
+                            buffer.usage,
+                            wgt::BufferUsages::INDIRECT,
+                        )
+                        .map_pass_err(scope)?;
+
+                        if offset % 4 != 0 {
+                            return Err(ComputePassErrorInner::UnalignedIndirectBufferOffset(
+                                offset,
+                            ))
+                            .map_pass_err(scope);
+                        }
+
+                        let end_offset =
+                            offset + mem::size_of::<wgt::DispatchIndirectArgs>() as u64;
+                        if end_offset > buffer.size {
+                            return Err(ComputePassErrorInner::IndirectBufferOverrun {
+                                offset,
+                                end_offset,
+                                buffer_size: buffer.size,
+                            })
+                            .map_pass_err(scope);
+                        }
+
+                        let (validation_pipeline_id, validation_bgl_id) = self
+                            .device_get_or_create_dispatch_indirect_validation_pipeline::<A>(
+                                device_id,
+                            )
+                            .map_err(ComputePassIndirectValidationError::ValidationPipeline)
+                            .map_pass_err(scope)?;
+
+                        let (dst_buffer_id, error) = self.device_create_buffer::<A>(
+                            device_id,
+                            &crate::resource::BufferDescriptor {
+                                label: None,
+                                size: 4 * 3,
+                                usage: wgt::BufferUsages::INDIRECT | wgt::BufferUsages::STORAGE,
+                                mapped_at_creation: false,
+                            },
+                            None,
+                        );
```

Comment:
Creating a buffer and bind group every dispatch is going to be _very_ expensive. In this case we can make a single buffer/bind group per queue and a single bind group per indirect command buffer. We'd trade one bind group for two, but mean we never have to create resources at runtime.

I really wish we had root-uavs/bda so we didn't need a bind group here.

Reply:
Wouldn't a single dst buffer (with its own bind group) per queue become a bottleneck in cases where a user issues 2 `dispatchIndirect` commands back to back each with their own indirect buffer? By only having one dst buffer, the 2 internal `dispatche`s we do for validation can't run in parallel.

Reply:
We discussed this at our maintainers meeting and decided to go with Connor's suggestion. My concerns about not being able to run dispatches in parallel were not well-founded as most (if not all) IHVs will treat buffer barriers as "wait for idle" anyway.